### ISSUE TITLE
UnixProcess#getInputStream returns InputStream that blocks reading untill some buffer is filled

### ISSUE THREAD

**Issue by allter** [scala-native/scala-native#4023]
scala-native versions: 0.4.17 , 0.5.4

A code for reproduction:

```
object Main {
  def main(args: Array[String]): Unit = {
    val p =
      new ProcessBuilder("bash", "-c", "while true; do date; sleep 1; done")
        .start()
    Thread.sleep(200)
    val buffer = new Array[Byte](8192)
    val n = p.getInputStream.read(buffer)
    println(s"read $n bytes: [${new String(buffer)}]")
  }
}
```

On JVM this (scastie: https://scastie.scala-lang.org/nNob35TPT0ewaB7qmatMaw ) immediately prints something like this:
```
read 29 bytes: [Thu Aug 15 12:39:15 UTC 2024
]
```

On Scala-Native this waits until all the buffer is filled:
```
read 8192 bytes: [Thu Aug 15 12:43:15 UTC 2024
Thu Aug 15 12:43:16 UTC 2024
...
Thu Aug 15 12:49:17 UTC 2024
Thu Aug ]
```

It looks like this is because some layer of buffering is added to javalib implementation of UnixProcess:

```
  class StreamImpl(val process: GenericProcess, is: FileInputStream)
      extends BufferedInputStream(is)
      with Stream {
```

**Comment by LeeTibbert** [comment]
allter

Thank you for the report and reproduction information.

Forgive me if this ***** in a hurry missed, but what operating system are you using? 
Thank you.
 
The reason that I ask is that `UnixProcess` actually has two implementations, Windows, 32 bit,
 and ones I could not exercise take the historical, path while Linux, macOS, and possibly other unix-like
operating systems take another. The latter itself has at least two paths: `epoll()` and `kexec()`.

`val n = p.getInputStream.read(buffer)` 

I'll have to re-read the relevant JVM documentation to refresh my understanding of the `read()` semantics.
There is specification and then there is as-implemented-in-the-real-world.

Off the top of my head, I would expect the `read()` to return on the first newline (beware os differences) or
 EOF. However, we are talking pipes here and they are their entirely own fiefdom.

**Comment by LeeTibbert** [comment]
allter.   

Another question, please,  to help me orient myself.  What version of Java are you
using with scastie?  I expect there to be no difference, but always nice to know
to keep in the back of ones mind when trying to replicate.

After a quick run thru the Java docs and the SN code base whilst waiting to compile:
It looks like the `StreamImpl` snippet you kindly provided is in `java.lang.process.PipeIO.scala`.

Java 22 is much clearer about distinguishing what is supposed to be buffered and what 
is not. It and Java 8 imply but do not state (or at least not that I could find in time stolen from
other projects) that `getInputStream()` is un-buffered, as you experience & report. `BufferedReader`
is provided in Java ??.   The Java docs state that, in my words not theirs, it is wise for the user
to buffer the stream from `getInputStream()`.

What this means is that I concur that this is a bug and will mark this report as such.

`PipeIo` should be extending simple, unbuffered `InputStream`.  Now it will take somebody
doing experiments to tell if that is a `PipedInputStream` cast to its super class `InputStream`.

To my thinking, any fix would be a breaking fix and would, skipping lightly over any resource issues,
have to be delayed until SN 0.6.n.

Do you have a workaround until then?  

Something like one of the `readLines()` methods instead of reading into a too large
 buffer might be more reliable, if they could be gotten to work.  A much better protocol
than "read & return what you have", but still not perfect.

**Comment by LeeTibbert** [comment]
@armanbilge   

How are your projects affected by a delayed solution here?  Has this bug shown up
in them or for your users?

**Comment by allter** [comment]
>  what operating system are you using

I'm using Ubuntu, kernel 5.13.0-52-generic, x86_64
I'm executing sbt using nix (for example, using scripts/nix-run from scala-native source code) and **plus** I've also tried linking statically using musl:
```
  pkgsCross = pkgs.pkgsCross.musl64;
  stdenv = pkgsCross.clangStdenv;
  jdk = pkgs.jdk8;
  clangPath =  stdenv.cc + "/bin/x86_64-unknown-linux-musl-clang";
  clangPPPath = stdenv.cc + "/bin/x86_64-unknown-linux-musl-clang++";
```
So this is not a particualar libc issue.

> UnixProcess actually has two implementations

I think my implementation is UnixProcessGen2, although I think that this issue is also both for UnixProcessGen1 and for WindowsProcess because all of them use PipeIO[PipeIO.Stream] which seem to add a buffering layer

> Off the top of my head, I would expect the read() to return on the first newline

IIRC that JVM uses system **read**, which doesn't take newlines into an account. System libc's **read** either:
1. If there is something in a file descriptor, get it up to the available buffer size (but if there are less than a buffer size, return what is available, even a single byte)
2. If there is EOF, return 0 bytes
3. If there is nothing in a file descriptor, block OS thread until at least one byte is available or EOF happens
4. If there is an error, return -1 as a size and set **errno**

**Comment by allter** [comment]
> The Java docs state that, in my words not theirs, it is wise for the user to buffer the stream from getInputStream().

This is wise for any program that grinds **a lot** of data from other processes because the program processes data in large chunks. But my use case is to try using scala-native to write a glue code that processes data from other unix programs. With the input buffering that a programmer is not able to turn off, it is impossible to communicate with a some kind of interactive programs.

> any fix would be a breaking fix and would, skipping lightly over any resource issues,
have to be delayed until SN 0.6.n.

I agree that some programs may rely on a current behaviour and may slow down or even break if badly designed. 
For me, I'd like the fix to be backported to 0.4.x so I can continue to try SN with cats-effect stack until they finally support 0.5 and later versions of SN.

> Do you have a workaround until then?

As for me, I am just trying SN with cats-effect and fs2 in my pet project, so I'm not greatly affected.

**Comment by ekrich** [comment]
> IIRC that JVM uses system read, which doesn't take newlines into an account. System libc's read either:

Technically, we are not suppose to view Java's GPL code as this project uses the Apache license so we can't infringe on copyleft. We can run tests etc. to see behaviors but shouldn't view code and then contribute.

**Comment by armanbilge** [comment]
> How are your projects affected by a delayed solution here? Has this bug shown up
> in them or for your users?

Yes, this was originally reported as https://github.com/typelevel/fs2/issues/3462.

> For me, I'd like the fix to be backported to 0.4.x so I can continue to try SN with cats-effect stack until they finally support 0.5 and later versions of SN.

I would say the burden is on us (Typelevel) to upgrade Cats Effect to SN 0.5 (instead of placing backport burden on the SN project, as the 0.4 series is effectively end-of-life). I'm currently working towards this.

---

> To my thinking, any fix would be a breaking fix and would, skipping lightly over any resource issues,
have to be delayed until SN 0.6.n.

@LeeTibbert thanks for your analysis, and permit me to offer a different perspective. In general, I agree that we want to provide compatibility guarantees for Scala Native APIs. However, for the special case of the JDK APIs, our goal is not to be compatible with previous (bugged) implementations provided by Scala Native. Our goal is to be compatible with the JDK/JVM specification and implementation.

In any case, if Scala Native is unable to fix this, we will fix it directly in FS2 by bypassing the JDK and directly calling the underlying APIs. I already had plans to do this anyway so that we can support non-blocking I/O.

**Comment by LeeTibbert** [comment]
Thank for all the info.  End of session, so sorry to be terse.

I think the next step at the SN end is get an idea of the size of the problem, the fix, and first
order ramifications.  I think that means first some private sandbox work to return not only
an InputStream (abstract), but a concrete implementation of the InputStream. Rabbit hole #1.

Then the trick is to see how SN CI reacts in a draft PR. What breaks when the new set of gears
grind.  ProcessTest has a habit of glitching whenever the direction of the wind changes.

**Comment by LeeTibbert** [comment]
I set up a private sandbox and tried some things. Who needs food & sleep anyway?

An obvious fix did not work, and things went downhill from there.  Looks like
I will have to do more tracing & experimentation, when I can find the time.

**Comment by LeeTibbert** [comment]
I am worked my way through enough of my backlog that I can look at this again.  Hurricane 
season was brutal-adjacent.

I want to study a solution where the `read(buf, off, len)` override of a `PipeIO` which does not
extend `BufferedInputStream` would examine the number of bytes available in the java fd,
which chains to the OS fd.  If zero bytes were available, it would block waiting for one byte.
Otherwise, it would use the min of the len & available (here len has been checked to be
strictly positive).   

I think this matches JDK 8 thru current 23 descriptions for bulk read. Namely, that
the method can return less than the requested len bytes.   

I think it also matches how processes generate output.

The trick will be if, say,  a child process blocks excessively because the parent is not reading the
output rapidly enough. 

I hope I enjoyed whatever evil living brought me to this juncture.
